File: dsa_leetcode_beginner/0091-decode-ways/0091-decode-ways.py
class Solution:
    def numDecodings(self, s: str) -> int:
        if not s or s[0] == '0':
            return 0

        n = len(s)
        dp = [0] * (n + 1)

        # Base cases
        dp[0], dp[1] = 1, 1  

        for i in range(2, n + 1):
            # Single digit decodec
            if 1 <= int(s[i-1:i]) <= 9:
                dp[i] += dp[i-1]

            # Two digit decode
            if 10 <= int(s[i-2:i]) <= 26:
                dp[i] += dp[i-2]

        return dp[n]
    
    
#Question: https://leetcode.com/problems/decode-ways
#Blog: https://blog.unwiredlearning.com/decode-ways

File: dsa_leetcode_beginner/0091-decode-ways/0091-decode-ways.cpp

#include <string>
#include <vector>

class Solution {
public:
    int numDecodings(const std::string &s) {
        if (s.empty() || s[0] == '0') {
            return 0;
        }

        int n = s.length();
        std::vector<int> dp(n + 1, 0);

        // Base cases
        dp[0] = 1;
        dp[1] = 1;

        for (int i = 2; i <= n; i++) {
            // Single digit decode
            if (s[i - 1] != '0') {
                dp[i] += dp[i - 1];
            }
            // Two digit decode
            int twoDigit = std::stoi(s.substr(i - 2, 2));
            if (twoDigit >= 10 && twoDigit <= 26) {
                dp[i] += dp[i - 2];
            }
        }

        return dp[n];
    }
};
